docs/design/rosbag2_playback_time.md

```diff
@@ -0,0 +1,200 @@
+   * clock's rate internally, so that the caller does not have to save it in order to resume.
+   */
+  void set_paused(bool paused);
+  bool get_paused() const;
+
+  /**
+    * Changes the rate of playback. Defaults to 1.0
+    * \param rate: must be nonzero positive. To pause playback, use \sa set_paused instead
+    * \throws std::invalid_argument if rate is <= 0.0
+    */
+  void set_rate(float rate);
+  float get_rate() const;
+
+  /**
+   * Set the rate that the clock will be published. Defaults to 0.
+   * \param frequency If this is set to <= 0, then the clock will not be published
+   */
+  void set_clock_publish_frequency(float frequency);
+  float get_clock_publish_frequency() const;
+
+  /**
+    * Change the current internally maintained offset so that next published time is different.
+    *
+    * This will trigger any registered JumpHandler callbacks.
+    * Call this with the first message timestamp for a bag before starting playback (otherwise this will return current wall time)
+    */
+  void jump(rclcpp::Time time);
+
+  rclcpp::JumpHandler::SharedPtr
+  create_jump_callback(
+    rclcpp::JumpHandler::pre_callback_t pre_callback,
+    rclcpp::JumpHandler::post_callback_t post_callback,
+    const rcl_jump_threshold_t & threshold);
+```
+
+### Clock Rate and Time Extrapolation
+
+This section is only relevant for Case 3 (subscribing to an external ROS Time Source), and does not apply to other cases.
+
+Common `/clock` publish rates are 10-100Hz - which is considerably slower than, for example, an IMU which may commonly publish at 200Hz or faster.
+According to the [ROS Time Source Design](http://design.ros2.org/articles/clock_and_time.html#ros-time-source), "calls to the ROS time abstraction will return the latest time received from the `/clock` topic."
+This means that faster topics will have a playback resolution only as fine as the `/clock` rate, as illustrated below.
+
+```
+# dots are arbitrary time tick, letters are when a message is published
+# M is topic Message, C is /clock message
+
+# original system
+M..M..M..M..M..M..M..M..M..M..M..
+C.......C........C........C......
+
+# playback on use_sim_time with default behavior - when a new time sample is received a burst of backed up messages will be played
+M.......MM.......MMM......MMM....
+C.......C........C........C......
+```
+
+The next section "No Advanced Estimating Clock By Default" lays out that more advanced behavior is possible, though "these techniques will require making assumptions about the future behavior of the time abstraction. And in the case that playback or simulation is instantaneously paused, it will break any of these assumptions."
+
+While the above is technically true, I propose that we implement a simple extrapolation anyways for the case of subscribing to an external ROS Time Source:
+* Store internally the last 2 time samples (noting the rate that samples are received)
```

Comment:
With using only two samples to determine rate this could be a little noisy.  but thats probably fine.